aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -775,23 +781,423 @@ Tensor bmm_nested(const Tensor& self, const Tensor& mat2) {
+    // determine the effect of splitting implicit batch dimension:
+    // 1. a new leading dimension with size `n2stack` and stride `Δoffsets`
+    // 2. offsets correspond to stacked tensors are merged to the head
+    ndims_underlying++;
+    std::vector<IntArrayRef> sizes_splitted, strides_splitted;
+    std::vector<int64_t> offsets_splitted(ntensors_reshaped);
+    std::vector<std::vector<int64_t>> sizes_splitted_vector(ntensors_reshaped),
+        strides_splitted_vector(ntensors_reshaped);
+    for (int64_t itensor_reshaped = 0; itensor_reshaped < ntensors_reshaped; itensor_reshaped++) {
+      int64_t itensor = itensor_reshaped * n2stack;
+      const IntArrayRef& size = sizes[itensor],
+          & stride = strides[itensor];
+      // we have guaranteed ntensors > ntensors_reshaped > 0,
+      // so ntensors >= 2, n2stack >= 2
+      for (int64_t i = 1; i < n2stack; i++) {
+        TORCH_CHECK(
+          size == sizes[itensor + i],
+          "cannot stack underlying tensors because they have different sizes");
+        TORCH_CHECK(
+          stride == strides[itensor + i],
+          "cannot stack underlying tensors because they have different strides");
+      }
+      offsets_splitted[itensor_reshaped] = offsets[itensor];
+      std::vector<int64_t>& size_splitted = sizes_splitted_vector[itensor_reshaped],
+          & stride_splitted = strides_splitted_vector[itensor_reshaped];
+      size_splitted.resize(ndims_underlying);
+      stride_splitted.resize(ndims_underlying);
+      size_splitted[0] = n2stack;
+      stride_splitted[0] = offsets[itensor + 1] - offsets[itensor];
+      for (int64_t i = 2; i < n2stack; i++) {
+        TORCH_CHECK(
+          stride_splitted[0] == offsets[itensor + i] - offsets[itensor + i - 1],
+          "cannot stack underlying tensors because they are not evenly spaced");
+      }
+      for (int64_t idim = 1; idim < ndims_underlying; idim++) {
+        size_splitted[idim] = size[idim - 1];
+        stride_splitted[idim] = stride[idim - 1];
+      }
+      sizes_splitted.push_back(IntArrayRef(size_splitted));
+      strides_splitted.push_back(IntArrayRef(stride_splitted));
+    }
+    // reshaping underlying tensor dimensions does not change offset further
+    // determine reshaped size and stride
+    bool reshape_as_view;
+    Tensor sizemat_reshaped, stridemat_reshaped;
+    std::tie(reshape_as_view, sizemat_reshaped, stridemat_reshaped) = NestedTensor_reshape_size_stride(
+        sizes_splitted, strides_splitted, proposed_shape, sizemat.options());
+    if (reshape_as_view) {
+      return wrap_buffer(buffer, sizemat_reshaped, stridemat_reshaped, offsets_splitted);
+    }
+    else {
+      Tensor buffer_reshaped = buffer.new_empty(buffer.sizes());
+      Tensor output = wrap_buffer(buffer_reshaped, sizemat_reshaped);
+      NestedTensor_reshape_copy(output,
+          buffer, sizes_splitted, strides_splitted, offsets_splitted);
+      return output;
+    }
```

Comment:
i see this is copy / pasted below as well - consider a utility that encapsulates the "if view do this, if copy do this" reshape logic